common/g3sinksUtils.hpp

```diff
@@ -0,0 +1,57 @@
+/** ==========================================================================
+* 2020 Joel Stienlet
+* this file is part of g3log by KjellKod.cc 
+*
+* This code is PUBLIC DOMAIN to use at your own risk and comes
+* with no warranties. This code is yours to share, use and modify with no
+* strings attached and no restrictions or obligations.
+* ============================================================================*/
+
+
+#pragma once
+
+/*
+ 
+ These deep copy types have been introduced to make it easier to
+ call sink methods with dynamically allocated data, without 
+ having to wait for the worker thread to finish.
+ They work together with the "NoJn"(No thread Join)-suffixed variants of
+ the corresponding sink methods.
+ 
+ Alternatively, you can:
+  - use static data
+  - wait for the thread (join it) with std::future .get()
+
+ */
+#include <cstring>
+#include <functional>
+#include <memory>
+
+namespace g3
+{
+
+// RAII string type with its deleter (Dltr) included to cross library boundaries 
+class DltrStr
+{
+public:
+    DltrStr() = delete;
+    // the deleter accompanies the memory block to enable it beeing freed in the library (using a different heap, as on Windows):
+    DltrStr(const char * content) {
+        size_t len = strlen(content);
+        _content = std::shared_ptr<char>(new char[len + 1], [](char *b){ delete[](b);} );
+        memcpy(_content.get(), content, len+1); };
+        
+    DltrStr(const std::string& content) {
+        size_t len = content.length();
+        _content = std::shared_ptr<char>(new char[len + 1], [](char *b){ delete[](b);} );
```

Comment:
I'm honestly not sure about this approach. Do you have some references/documentation that will back this up?

Reply:
L46 - 47

Reply:
If this is correct (I've to read up on this too) then this class can be greatly simplified.   Please verify this however. https://stackoverflow.com/questions/345003/is-it-safe-to-use-stl-tr1-shared-ptrs-between-modules-exes-and-dlls

Reply:
Interesting read this discussion on Stackoverflow.
Marc Mutz says that a shared_ptr is cross DLL-safe by design, because "there's no such thing as a shared_ptr without a deleter".
So then the deleter could be removed in the class I suggested. Or even remove that class altogether, and pass directly the shared_ptr as argument. But I have some preference for the class, as it allows making changes to the allocation without touching the interface.

But I think there could still be a problem:
let's suppose we have an exe, g3sinks, and another_lib that uses g3sinks.
Let's suppose the exe calls a function from "another_lib" that ends up creating a DltrStr sent to g3sinks. But now suppose that, before the thread from g3log has finished using and deleting that DltrStr, another_lib's dll gets unloaded. Then the deleter inserted into DltrStr (by another_lib) and called in g3sinks' thread would no longer be valid.
This may be a corner case, but I think the documentation should warn about premature unloading of a dll.
All threads from g3sinks must be joined before a lib using g3log gets unloaded.

Reply:
I wouldn’t jump on what Marc Mutz said without verifying it in the standard. He might be right or wrong. 

There’s a dll loading example in G3log that you can always play with to verify there’s no memory corruption

Reply:
I add another reference:
From Effective C++ Digital Collection: 140 Ways to Improve Your Programming
by Scott Meyers
 "An especially nice feature of tr1::shared_ptr is that it automatically uses its per-pointer deleter to eliminate another potential client error, the "cross-DLL problem."
https://books.google.fr/books?id=U7lTySXdFk0C&lpg=PP1&pg=PT149#v=onepage&q&f=false

Reply:
Your concern about DltrStr and unloading dll is valid. 
However the recommendation that g3sinks should be joined before dlls are shutting down. I.e G3log is in exit mode after hitting the RAII destruction sequence may be contrary to best practices for G3log. 

It’s assumed that G3log exits (RAII) very late in the shutdown sequence so that any shutdown logging is captured. 

Personally all of this can be avoided by following the recommendation of waiting for the returned future and by not sending a temporary (that expires) to the sink. 
The `call(...) ` was never intended as a high performance frequent calling API but as a thread safe (following the practices we already discussed) way to reach any public function on the sinks API

Reply:
I guess another way could be to skip the DltStr object but keep the added function to the SyslogSink but having it receive a `shared_ptr` instead. 

The theoretical dll client using the shared_ptr API on the sink is now made aware (or should be) that he shouldn’t shut down before the call is done executing.

Reply:
For a shutdown sequence the client is always responsible to let future’s etc be done (hard to avoid with a returned future, right?)  so the only thing to worry about is to not sending in a temporary ... which is common thread data passing practice

Reply:
I totally agree that we better test this before adding it to g3sinks. Can I use MinGW for this? (I almost never use Windows, so I've no idea about the right toolchains to use)

Reply:
I'm thinking I should have explained the purpose earlier: the difficulty I had to begin with was for making a python wrapper (to use the same logger in both python and c++):
https://github.com/JoelStienlet/g3logPython
using pybind11.
I could join the threads in the c++ part, but then it blocks the python code until the thread finished, and the python code cannot decide when to join the thread.
Or
I could add these DltStr strings, and then I can join the threads from the python code without worrying for the garbage collector to delete the string in the meantime. (what I observed, and was causing crashes)

Reply:
I’ll check it out. Very cool usage btw. 

Here’s a dll loading test for nix systems. There’s another one for Windows in the same directory if you need it. 

https://github.com/KjellKod/g3log/blob/master/test_unit/test_linux_dynamic_loaded_sharedlib.cpp

Reply:
Btw: you AFAIK you don’t need memcopy for the shared_ptr and so you can greatly simplify the code.  ... unless I missed something

Reply:
I'm thinking, perhaps it would be better to manage the lifetime of the strings entirely in the wrapper, and don't change your library at all? I will have to store the allocated strings along the corresponding futures, and "regularly" check if any thread has finished. I don't see any reason why this couldn't be done, so perhaps this is the best way to go after all.